sessionctx/variable/session.go

```diff
@@ -1907,3 +1923,85 @@ func (s *SessionVars) SlowLogFormat(logItems *SlowQueryLogItems) string {
 func writeSlowLogItem(buf *bytes.Buffer, key, value string) {
 	buf.WriteString(SlowLogRowPrefixStr + key + SlowLogSpaceMarkStr + value + "\n")
 }
+
+func (s *SessionVars) ResetTxnTimes() {
+	s.StartTimes = s.StartTimes[:0]
+	s.EndTimes = s.EndTimes[:0]
+}
+
+// ReportTxnAndLockDuration reports txn and the pessimistic locks' idle and duration to metric server.
+// Run this function in another goroutine to avoid it blocks the Commit or Rollback statements.
+func (s *SessionVars) ReportTxnAndLockDuration(startTS uint64, keysArr []kv.LockedKeys) {
+	if len(s.EndTimes) == 0 {
+		return
+	}
+	// copy the slice because it may be cleaned in other goroutines
+	startTimes := make([]time.Time, len(s.StartTimes))
+	copy(startTimes, s.StartTimes)
+	endTimes := make([]time.Time, len(s.EndTimes))
+	copy(endTimes, s.EndTimes)
+	// the rest works can be executed in another goroutine, so that it won't block the session goroutine.
+	go func() {
+		if len(startTimes) != len(endTimes)+1 {
+			panic(fmt.Sprintf("incorrect log durations, startTimes: %v, endTimes: %v", startTimes, endTimes))
```

Comment:
It will be used directly in the serious production environment, better not to panic.

Reply:
I'll remove it, log an error and skip the duration calculation.